oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst

```diff
@@ -0,0 +1,65 @@
+To Make or Not to Make
+======================
+
+
+Status
+------
+
+Accepted
+
+
+Context
+-------
+
+For Decentralized Devstack, an common interface is necessary to enable a OpenedX Developer to add contributions to the many independent services in Open edX codebase.
+
+Options
+~~~~~~~
+
+1. The current Devstack uses make targets as an basic interface for developers.
+
+2. Use docker-compose commands directly
+
+Case for and against make targets
+`````````````````````````````````
+
+For:
+
+- This standard already exists and is known by developers at edx.
+
+  * If we don't create a make interface now, sevices might create their on make interface, which might/will result in non-standard make experience
+- make targets decrease verbosity and memorization
```

Comment:
> Enable a really consistent UI across projects based on different tech stacks (although we don't currently enforce this to the degree I think we should)

IMO, this is a non-goal. DD is largely built on the decisions made in OEP-41, whose purpose is to standardize back-end services on single tech stack. Furthermore, I'm generally against trying to build abstractions on top of tools that are already meant to be end-user-facing abstractions.

> Provide a "canonical example" of how to do those operations

+1, a well-curated Makefile does have a way of serving as supplemental documentation. I would list this as a pro. However, we'll need to write docs anyway, whether it be for `make` commands, `docker-compose` commands, or a mix of both.

> Reduce the onboarding learning curve by that consistency and these examples

A Makefile isn't a necessary nor sufficient condition for a consistent interface. Some Makefiles are well-curated, others  become inconsistent messes. Some docker-compose projects are non-idiomatic and confusing to use, some are well-maintained and a breeze to use if you know the four or five basic `docker-compose` commands.

I say this because I want to dispel the notion that a decision to use docker-compose directly is a decision to throw a consistent interface out the window.

Reply:
> MO, this is a non-goal. DD is largely built on the decisions made in OEP-41, whose purpose is to standardize back-end services on single tech stack

A usable development environment consists of more than just back-end services and that trend is accelerating

Reply:
> I say this because I want to dispel the notion that a decision to use docker-compose directly is a decision to throw a consistent interface out the window.

Also, for sure, our current interface is not as consistent as it should be right now.

Using docker-compose directly _can_ be consistent as long as _everything_ uses Docker; if some things (like frontend projects) work differently, wrappers can smooth those differences and make the developer experience more consistent.

Reply:
> Using docker-compose directly can be consistent as long as everything uses Docker; if some things (like frontend projects) work differently, wrappers can smooth those differences and make the developer experience more consistent.

I think this teases at the root of our disagreement, @inventhouse .

I would prefer that front-ends standardize on one thing (e.g. NPM), back-end services standardize on another (e.g. Docker+Compose), and developers learn both `npm` and `docker-compose`. They seem like different tools for different jobs, and wrapping them may just obscure & weaken the power of the underlying tools.

But I recognize your desire to factor out the common use cases between the different tools, such as up/down/restart. What I wonder is: how would you address operations in one paradigm that have no analog in the other? For example, you can't "do database migrations" a micro-frontend. So would you implement that as a no-op for frontends? Or would that be left out of the common interface?

Reply:
Is there already a working prototype of a minimal decentralized devstack running somewhere? E.g: something that would include just the LMS, CMS, forum, an MFE and the data layer services. My feeling is that having a prototype would really help providing an immediate answer to these conversations.

As far as I am concerned, I fail to understand how such a decentralized devstack will work. For instance: would the user have to declare one or multiple docker-compose.yml files? If it is the latter, how does one automate passing the (possibly) long list of docker-compose.yml files on the CLI? How to configure the different IDAs when optional services may not be running at all? (e.g: what is the value of `COMMENTS_SERVICE_URL` when the forums may or may not be running?)

I understand that this OEP is not the right place to answer these questions; but having a working prototype would help clarify many of the technical issues (IMHO).

Reply:
Hi @edx/core-committer-regisb here is link to PR with working prototype: https://github.com/edx/enterprise-catalog/pull/195. It still needs to be tested, but your feedback on this would be welcome

Reply:
you can find documentation for it in docs folder

Reply:
This is a great example @jinder1s! I see that there are already make commands there. For instance:

    dev.migrate: # Migrates databases. Application and DB server must be up for this to work.
	docker exec -it enterprise.catalog.app bash -c 'cd /edx/app/enterprise_catalog/enterprise_catalog && make migrate'

If we were to replace this command by a docker-compose command, how would we do it? For instance we could run:

    docker-compose run --rm -f decentralized-docker-compose.yml enterprise.catalog.app make migrate

Would this be acceptable? Basically, this moves the `make` commands from the devstack to the service running inside the container.

(BTW: I strongly recommend *not* to use `docker exec`. In my experience with Tutor, this was causing major issues:

1. It requires the platform to be already running. This might not be the case because the container might be crashing because the data services are not properly configured.
2. Environment variables defined in docker-compose.yml or the .env file will not be taken into account

Sorry for the long parenthesis. If you'd like I can move this comment to the enterprise-catalog PR)

Reply:
for that particular question, I tried documenting here: [doc link](https://github.com/edx/enterprise-catalog/blob/msingh/spike/merge_ready/docs/decentralized_devstack_faq.rst). 

FYI, we are using .env file to toggle between decentralized devstack and legacy devstack as documented here: [toggling documentation](https://github.com/edx/enterprise-catalog/blob/msingh/spike/merge_ready/docs/decentralized_devstack_workflows.rst#toggle-between-decentralized-devstack-and-legacy-devstack)

Reply:
Can you elaborate on your recommendation against using docker exec? What alternative would you recommend? Are you suggesting using docker-compose exec as an alternative or wrapping docker exec in make target?

Reply:
> Can you elaborate on your recommendation against using docker exec? What alternative would you recommend?

In my experience, `docker run --rm -it` is better than `docker exec -it`, and `docker-compose run --rm` is even better.